### PR TITLE
SAK-22892 Users can now be added to sites via JSON

### PR DESCRIPTION
**Pull request by adrianfish**

### CODE REVIEW

**Comment by adrianfish** [comment]
We've had this in production for around 18 months now and it's been fine. I won't merge as this was Aaron's turf and I'm still deeply respectful of it.

File: entitybroker/core-providers/pom.xml [hunk 1]
```diff
@@ -54,6 +54,16 @@
             <artifactId>mail</artifactId>
         </dependency>
         <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
```
Comment (ottenhoff) [inline entitybroker/core-providers/pom.xml]: Note that SAK-28977 added Jackson libs to shared/lib

File: entitybroker/core-providers/src/java/org/sakaiproject/entitybroker/providers/MembershipEntityProvider.java [hunk 2]
```diff
@@ -309,6 +316,113 @@ public ActionReturn handleSiteMemberships(EntityView view, Map<String, Object> p
...
+        // Collect the users into roles
+        Map<String, List<String>> usersToRoleMap = new HashMap<String, List<String>>();
+        for (JsonUser user : memberships) {
+            String id = user.getId();
+            String role = user.getRole();
+            
+            if (usersToRoleMap.containsKey(role)) {
+                usersToRoleMap.get(role).add(id);
+            } else {
+                List<String> users = new ArrayList<String>(20); // Prealloc space for 20 users in each role
```
Comment (steveswinsburg) [inline entitybroker/core-providers/src/java/org/sakaiproject/entitybroker/providers/MembershipEntityProvider.java]: Not an issue, but this doesn't really save a lot in terms of reallocation costs as we can't really guess how big the list of users is going to be: could be 20, could be 2000. Might even be just 1 or 2 in which case this is wasting (a tiny bit of) memory.

File: entitybroker/core-providers/src/java/org/sakaiproject/entitybroker/providers/MembershipEntityProvider.java [hunk 3]
```diff
@@ -309,6 +316,113 @@ public ActionReturn handleSiteMemberships(EntityView view, Map<String, Object> p
...
+                users.add(id);
+                usersToRoleMap.put(role, users);
+            }
+        }
+		
+        Map<String, Object> localParams = new HashMap<String, Object>(2);
```
Comment (steveswinsburg) [inline entitybroker/core-providers/src/java/org/sakaiproject/entitybroker/providers/MembershipEntityProvider.java]: Likewise. IMO the numbers can be confusing, I just avoid preallocation altogether.

**Comment by steveswinsburg** [comment]
I'm pretty happy with this, though I haven't pulled the branch to test. The only thing I would change is the preallocation numbers just to make it that tiny bit clearer but that is a personal taste.

**Pull request merged**